### PR TITLE
add info/error messages for SSD/HDD support - LSO mode

### PR DESCRIPTION
**Pull request by SanjalKatiyar**
https://issues.redhat.com/browse/RHSTOR-5835
https://issues.redhat.com/browse/RHSTOR-5834

#### SSDs are detected (not blocking users from deployment):
Screenshot 2024-06-19 at 1 40 25 PM

#### No SSDs are detected (blocking users from deployment):
Screenshot 2024-06-19 at 1 40 48 PM

#### No SSDs detected (blocking users from deployment):
Screenshot 2024-06-19 at 1 44 43 PM

#### In case LSO is deleted or any other edge case (not blocking users from deployment, showing a general info message instead):
Screenshot 2024-06-19 at 1 47 12 PM

### CODE REVIEW

File: packages/odf/hooks/useLSODiskDiscovery.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import * as React from 'react';
```
Comment (SanjalKatiyar) [inline packages/odf/hooks/useLSODiskDiscovery.ts:1]: this file is mostly a refactor, moving existing functionality into a separate hook... only needs high-level review...

File: packages/odf/components/utils/lso.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import { LABEL_OPERATOR, LABEL_SELECTOR } from '@odf/core/constants';
```
Comment (SanjalKatiyar) [inline packages/odf/components/utils/lso.ts:1]: this file is mostly a refactor, moving existing common utils into a separate file... only needs high-level review...

File: packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx [hunk 3]
```diff
@@ -166,12 +167,6 @@ const SelectCapacityAndNodes: React.FC<SelectCapacityAndNodesProps> = ({
 
   const replicas = getReplicasFromSelectedNodes(nodes);
 
-  React.useEffect(() => {
-    // Reset pvCount that could have been set by another StorageClass.
-    dispatch({ type: 'capacityAndNodes/pvCount', payload: 0 });
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
-
```
Comment (SanjalKatiyar) [inline packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx:174]: I remember we added this as part of "Performance mode" epic, but on taking a second look I see that we don't need this.
https://github.com/red-hat-storage/odf-console/blob/master/packages/odf/components/create-storage-system/reducer.ts#L192 should reset the redux state.

File: packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx [hunk 4]
```diff
@@ -437,30 +451,29 @@ export const CapacityAndNodes: React.FC<CapacityAndNodesProps> = ({
...
+    deploymentMode,
+    volumeValidationType
   );
   const onProfileChange = React.useCallback(
     (profile) => onResourceProfileChange(dispatch)(profile),
     [dispatch]
   );
 
-  React.useEffect(() => {
-    // Reset selected nodes that could have been set by another StorageClass.
```
Comment (alfonsomthd) [inline packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx:448]: @SanjalKatiyar I remember you suggesting this reset for safety. Is it no longer needed?

File: packages/odf/components/create-storage-system/reducer.ts [hunk 5]
```diff
@@ -70,20 +73,22 @@ export const initialState: CreateStorageSystemState = {
...
     resourceProfile: ResourceProfile.Balanced,
+    volumeValidationType: VolumeTypeValidation.NONE,
   },
   createStorageClass: {},
   connectionDetails: {},
   createLocalVolumeSet: {
     volumeSetName: '',
     isValidDiskSize: true,
-    diskType: 'All',
+    // we only support SSD, other options (All/HDD) are disabled
```
Comment (alfonsomthd) [inline packages/odf/components/create-storage-system/reducer.ts:83]: Just confirming: this only applies to internal mode, right?

File: packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx (hunk 4 above)
Comment (SanjalKatiyar) [reply packages/odf/components/create-storage-system/create-storage-system-steps/capacity-and-nodes-step/capacity-and-nodes-step.tsx:448]: https://github.com/red-hat-storage/odf-console/pull/1437#discussion_r1645714216

File: packages/odf/components/create-storage-system/reducer.ts (hunk 5 above)
Comment (SanjalKatiyar) [reply packages/odf/components/create-storage-system/reducer.ts:83]: yes, internal with LSO to be specific...

File: packages/odf/hooks/useLSODiskDiscovery.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  } catch (loadError) {
+    if (loadError?.response?.status === 404) {
+      try {
+        await createLocalVolumeDiscovery(
+          nodeByHostNames,
+          namespace,
+          OCS_TOLERATION
+        );
+      } catch (createError) {
+        setError(createError.message);
```
Comment (alfonsomthd) [inline packages/odf/hooks/useLSODiskDiscovery.ts:34]: Could this `setError` be overriden by `setError(false)` in the `finally` block?

Comment (SanjalKatiyar) [reply packages/odf/hooks/useLSODiskDiscovery.ts:34]: good point, probably need to check/test this once... can u plz create a GH issue ??

this is an already existing functionality, I just refactored it a bit in this PR: https://github.com/red-hat-storage/odf-console/pull/1437/files/e53efdc6f8a4616a203f7a341e0171029d94e920#r1643974261

Comment (alfonsomthd) [reply packages/odf/hooks/useLSODiskDiscovery.ts:34]: Done.

File: packages/shared/src/dropdown/multiselectdropdown.tsx [hunk 7]
```diff
@@ -53,7 +53,7 @@ export const MultiSelectDropdown: React.FC<MultiSelectDropdownProps> = ({
     <Select
       variant={variant || SelectVariant.typeaheadMulti}
       aria-label={t('Select input')}
-      onToggle={() => setOpen((o) => !o)}
+      onToggle={(_event, expanded) => setOpen(() => expanded)}
```
Comment (alfonsomthd) [inline packages/shared/src/dropdown/multiselectdropdown.tsx:56]: Should the callback inside `setOpen` return `!expanded` ?

Comment (SanjalKatiyar) [reply packages/shared/src/dropdown/multiselectdropdown.tsx:56]: I am not entirely sure of the reason but that's buggy somehow. Doing so is effectively opening and closing the toggle back (before even human eye can perceive it)... callback is getting executed twice and second execution is reverting back the set value to `false`...

whereas the change which I did as part of this PR works as expected...

Comment (SanjalKatiyar) [reply packages/shared/src/dropdown/multiselectdropdown.tsx:56]: that works, I have tested it... `expanded` value corresponds to the immediate user click...

Comment (alfonsomthd) [reply packages/shared/src/dropdown/multiselectdropdown.tsx:56]: Thanks for confirming it works.

**Comment by alfonsomthd** [comment]
LGTM
@SanjalKatiyar If possible, please add some unit tests for the new hooks.
Delegating the tagging to @bipuladh @GowthamShanmugam

**Comment by SanjalKatiyar** [comment]
> LGTM @SanjalKatiyar If possible, please add some unit tests for the new hooks. Delegating the tagging to @bipuladh @GowthamShanmugam

make sense, will probably add it in a different PR though... it's already getting long due to lot of refactoring...

**Comment by alfonsomthd** [comment]
> LGTM @SanjalKatiyar If possible, please add some unit tests for the new hooks. Delegating the tagging to @bipuladh @GowthamShanmugam

@bipuladh :arrow_up: (just FYI)

**Pull request merged**